main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTimeTravelSpeedValidator.java

```diff
@@ -0,0 +1,362 @@
+    for (int i = 0; i < distancesKm.length; ++i) {
+      S2LatLng nextLatLng = getStopLatLng(stopTable, stopTimes.get(i + 1).stopId());
+      distancesKm[i] = S2Earth.getDistanceKm(currLatLng, nextLatLng);
+      currLatLng = nextLatLng;
+    }
+    return distancesKm;
+  }
+
+  /**
+   * Validates travel speed between far stops for all trips in a given line variant.
+   *
+   * <p>If there is a fast travel detected, then exactly one notice is issued for each trip.
+   */
+  private void validateFarStops(
+      List<List<GtfsStopTime>> trips,
+      double[] distancesKm,
+      double maxSpeedInKph,
+      NoticeContainer noticeContainer) {
+    final List<GtfsStopTime> stopTimes = trips.get(0);
+
+    for (int endIdx = 0; endIdx < stopTimes.size(); ++endIdx) {
+      final GtfsStopTime endStopTime = stopTimes.get(endIdx);
+      if (!endStopTime.hasArrivalTime()) {
+        continue;
+      }
+      final GtfsStop endStop = stopTable.byStopId(endStopTime.stopId());
+      // distanceToEndIdx stores the distance between stops endIdx and startIdx for all startIdx <
+      // endIdx, computed as the sum of straight-line distances between consecutive stops.
+      double distanceToEndIdx = 0;
+      for (int startIdx = endIdx - 1; startIdx >= 0; --startIdx) {
+        final GtfsStopTime startStopTime = stopTimes.get(startIdx);
+        distanceToEndIdx += distancesKm[startIdx];
+        if (!startStopTime.hasDepartureTime()) {
+          continue;
+        }
+        final double speedKph =
+            getSpeedKphBetweenStops(distanceToEndIdx, startStopTime, endStopTime);
+        if (speedKph <= maxSpeedInKph) {
+          continue;
+        }
+        GtfsStop startStop = stopTable.byStopId(startStopTime.stopId());
+        // Give a notice if they are too far apart.
+        if (distanceToEndIdx > MAX_DISTANCE_OVER_MAX_SPEED_IN_KMS) {
+          // Issue one notice per each trip.
+          for (List<GtfsStopTime> stopsOnTrip : trips) {
+            noticeContainer.addValidationNotice(
+                new FastTravelBetweenFarStopsNotice(
+                    tripTable.byTripId(stopsOnTrip.get(0).tripId()),
+                    stopsOnTrip.get(startIdx),
+                    startStop,
+                    stopsOnTrip.get(endIdx),
+                    endStop,
+                    speedKph,
+                    distanceToEndIdx));
+          }
+          // Report at most once for each trip.
+          return;
+        }
+      }
+    }
+  }
+
+  /**
+   * Validates travel speed between consecutive stops for all trips in a given line variant.
+   *
+   * <p>If there is a fast travel detected, then a separate notice is issued for each trip.
+   */
+  private void validateConsecutiveStops(
+      List<List<GtfsStopTime>> trips,
+      double[] distancesKm,
+      double maxSpeedInKph,
+      NoticeContainer noticeContainer) {
+    final List<GtfsStopTime> stopTimes = trips.get(0);
+
+    for (int i = 0; i < distancesKm.length; ++i) {
```

Comment:
If I understand this logic correctly, `validateConsecutiveStops()` won't validate GTFS without the `timepoint` column. In other words, if there are rows in stop_times.txt without times included (i.e., the rows with times are implicitly timepoints):

|  stop_sequence | arrival_time | departure_time |
|  ------ | ------ | ------ | 
1 | 00:00:00 | 00:00:00  |  
2 |  |  |  
3 |  |  |  
4 | 00:10:00 | 00:10:00  |  

...because it would simply skip over the empty rows and look for the next row with a departure time.

But, these cases would still be evaluated by `validateFarStops()`.

Is that right @aababilov?

If so it might be useful to include some explanation of this in the comments to explain why the two different rule implementations exist.

Related - https://github.com/google/transit/issues/61.

Reply:
`timepoint` column does not affect this function, but it indeed depends on presence of `arrival_time` and `departure_time`.

Two rules exist to catch different situations:

* too high speed between consequent stops
* too high speed between stops that are far away (more than 10 km apart)

This is already explained twice: as a comment for the notice classes and in `RULES.md`.